# Project Euler: Problem 1

If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

<p> It is tempting to try to solve this problem in a very simple way. A possible solution could look like this: </p>

In [9]:
sum([x for x in range(1000) if x%3 == 0 or x%5 == 0])

233168

The issue with the previous line of code is that the computational cost is high because the for loop evaluates the if conditions for each number. As an example, I tested the code with 1000000000 (one billion) and my computer printed the answer in 1m 20.5s. I will try to optimize the solution to reduce that time. Additionally, I would like to create a more general solution that will be able to take more than one number and that will be able to provide the sum of all the multiples of 3 or 5 below each input.

If I want to sum all the integers from 1 to a 100, I could simply take the tedious task of adding one number after the other. There is a tale that says that Carl Friedrich Gauss solved this problem when he was in primary school. The strategy to simplify the task is to observe the following:
$$
1 + 100 = 101 \\
2 + 99 = 101 \\
3 + 98 = 101 \\
…  \\
50 + 51 = 101  \\
$$
Therefore, the sum of all integers from 1 to a 100 can be expressed as 101*50 which is 5050. With this concept, the sum of consecutive integers can be generalized as follows:
$$ \sum_{1}^{n}i = \frac{n(n + 1)}{2} $$

To implement this concept in the solution of the problem, the sum of the ordered multiples of a number should be expressed as that number times a series. For example, for 3 up to 15:

$$ 3 + 6 + 9 + 12 + 15 = 3*(1 + 2 + 3 + 4 + 5) $$

The addition in the parenthesis can be solved with the equation that I presented earlier, where n equals 5. For each case, n has to be determined and this can be done by dividing 1000, or any other number, by 3 and 5. This results in the following:

$$ sum = \frac{n_{(3)}(n_{(3)} + 1)}{2} + \frac{n_{(5)}(n_{(5)} + 1)}{2} $$

The problem is that this solution is adding some multiples two times. For example, 15 is a multiple of 5 and 3 and both terms in the last expression are adding 15. However, there is a simple step to correct this and that is to subtract one time the multiples of 15.

$$ sum = \frac{n_{(3)}(n_{(3)} + 1)}{2} + \frac{n_{(5)}(n_{(5)} + 1)}{2} - \frac{n_{(15)}(n_{(15)} + 1)}{2} $$



In [2]:
# Solution

t = int(input('Please input the number of test cases: ').strip())

for _ in range(t):
    n = int(input().strip())
    
    n3 = (n - 1)//3
    n5 = (n - 1)//5
    n15 = (n - 1)//15
    
    val3 = 3*n3*(n3 + 1)//2
    val5 = 5*n5*(n5 + 1)//2
    val15 = 15*n15*(n15 + 1)//2
    
    print(val3 + val5 - val15)

Please input the number of test cases: 2
10
23
1000
233168


This solution takes 0.3s for one billion, and can accept multiple test cases. 

The sum of all the multiples of 3 or 5, below 1000, is 233168.